cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaRebalanceAssemblyOperator.java

```diff
@@ -0,0 +1,869 @@
+import io.strimzi.operator.cluster.model.StatusDiff;
+import io.strimzi.operator.cluster.operator.resource.cruisecontrol.CruiseControlApi;
+import io.strimzi.operator.cluster.operator.resource.cruisecontrol.CruiseControlApiImpl;
+import io.strimzi.operator.cluster.operator.resource.cruisecontrol.CruiseControlUserTaskResponse;
+import io.strimzi.operator.cluster.operator.resource.cruisecontrol.CruiseControlUserTaskStatus;
+import io.strimzi.operator.cluster.operator.resource.cruisecontrol.RebalanceOptions;
+import io.strimzi.operator.cluster.operator.resource.ResourceOperatorSupplier;
+import io.strimzi.operator.common.AbstractOperator;
+import io.strimzi.operator.common.Annotations;
+import io.strimzi.operator.common.Reconciliation;
+import io.strimzi.operator.common.Util;
+import io.strimzi.operator.common.model.Labels;
+import io.strimzi.operator.common.operator.resource.AbstractWatchableResourceOperator;
+import io.strimzi.operator.common.operator.resource.CrdOperator;
+import io.strimzi.operator.common.operator.resource.StatusUtils;
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import io.vertx.core.Vertx;
+import io.vertx.core.json.JsonObject;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+
+import static io.strimzi.operator.cluster.operator.resource.cruisecontrol.CruiseControlApi.CC_REST_API_SUMMARY;
+
+/**
+ * <p>Assembly operator for a "KafkaRebalance" assembly, which interacts with Cruise Control REST API</p>
+ *
+ * <p>
+ *     This operator takes care of the {@code KafkaRebalance} custom resources that a user can create in order
+ *     to interact with Cruise Control REST API and execute a cluster rebalancing.
+ *     A state machine is used for the rebalacing flow which is reflected in the {@code status} of the custom resource.
+ *
+ *     When a new {@code KafkaRebalance} custom resource is created, the operator sends a rebalance proposal
+ *     request to the Cruise Control REST API and moves to the {@code PendingProposal} state. It stays in this state
+ *     until a the rebalance proposal is ready, polling the related status on Cruise Control, and then finally moves
+ *     to the {@code ProposalReady} state. The status of the {@code KafkaRebalance} custom resource is updated with the
+ *     computed rebalance proposal so that the user can view it and making a decision to execute it or not.
+ *     For starting the actual rebalancing on the cluster, the user annotate the custom resource with
+ *     the {@code strimzi.io/rebalance=approve} annotation, triggering the operator to send a rebalance request to the
+ *     Cruise Control REST API in order to execute the rebalancing.
+ *     During the rebalancing, the operator state machine is in the {@code Rebalancing} state and it moves finally
+ *     to the {@code Ready} state when the rebalancing is done.
+ *
```

Comment:
```suggestion
 * <pre><code>
 *   User        Kube           Operator              CC
 *    | Create KR  |               |                   |
 *    |----------->|   Watch       |                   |
 *    |            |-------------->|   Proposal        |
 *    |            |               |------------------>|
 *    |            |               |   Poll            |
 *    |            |               |------------------>|
 *    |            |               |   Poll            |
 *    |            | Update Status |------------------>|
 *    |            |<--------------|                   |
 *    |            |   Watch       |                   |
 *    |            |-------------->|                   |
 *    | Get        |               |                   |
 *    |----------->|               |                   |
 *    |            |               |                   |
 *    | Approve    |               |                   |
 *    |----------->|  Watch        |                   |
 *    |            |-------------->|   Rebalance       |
 *    |            |               |------------------>|
 *    |            |               |   Poll            |
 *    |            |               |------------------>|
 *    |            |               |   Poll            |
 *    |            | Update Status |------------------>|
 *    |            |<--------------|                   |
 *    |            |   Watch       |                   |
 *    |            |-------------->|                   |
 *    | Get        |               |                   |
 *    |----------->|               |                   |
 * </code></pre>
```

Reply:
Do you really want me to put this asciiart in the doc? Why not having official doc for this? Isn't it too much for a source file?

Reply:
Sure, why not? Don't you think it summaries the basic flow in a readily understandable form? Much more easily understood than having to read paragraphs of text, imho. And I didn't know there could be _too much_ documentation in a place like this were only us developers are going to benefit from it.

Reply:
Ok, I added it.

Reply:
You might thank me in 6 months time when you're having to debug a problem and can't remember how this code works :slightly_smiling_face: